.github/workflows/ci.yml

```diff
@@ -88,36 +95,69 @@ jobs:
     outputs:
       os: ${{ steps.setup-os-matrix.outputs.os }}
 
-  build-and-test:
+  # Check that building the entire solution for all target frameworks works.
+  build:
+    # Do not run this job for pull requests where both branches are from the same repo.
+    # Jobs that depend on this one will be skipped too.
+    # This prevents duplicate CI runs for our own pull requests, whilst preserving the ability to
+    # run the CI for each branch push to a fork, and for each pull request originating from a fork.
+    if: github.event_name == 'push' || github.event.pull_request.head.repo.id != github.event.pull_request.base.repo.id
+    runs-on: windows-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Setup the latest .NET 5 SDK
+        uses: actions/setup-dotnet@v1
+        with:
+          dotnet-version: 5.0.x
+
+      - name: Build
+        run: dotnet build --configuration=Release -p:TreatWarningsAsErrors=true Src
+
+  # Run tests for all supported combinations of OS/library/framework.
+  test:
     needs: setup-os-matrix
     strategy:
       matrix:
         os: ${{ fromJson(needs.setup-os-matrix.outputs.os) }}
         library: [ILGPU, ILGPU.Algorithms]
+        framework: [netcoreapp3.1, net5.0]
+        include:
+          - os: windows-latest
+            library: ILGPU
+            framework: net47
+          - os: windows-latest
+            library: ILGPU.Algorithms
+            framework: net47
       fail-fast: false
     runs-on: ${{ matrix.os }}
     steps:
       - name: Checkout
         uses: actions/checkout@v2
-      - name: Setup .NET Core SDK
+
+      - name: Setup the latest .NET 5 SDK
         uses: actions/setup-dotnet@v1
         with:
-          dotnet-version: 3.1.407
-      - name: Pin .NET Core SDK
-        run: dotnet new globaljson --sdk-version 3.1.407
-      - name: Release Build
-        run: dotnet build --configuration=Release -p:TreatWarningsAsErrors=true Src
-      - name: CPU Tests
-        if: matrix.os != 'cuda'
-        run: dotnet test --logger GitHubActions --no-build --configuration=Release Src/${{ matrix.library }}.Tests.CPU
-        env:
-          ILGPU_CLEAN_TESTS: ${{ github.event_name == 'push' && (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/tags/v')) }}
-      - name: CUDA Tests
-        if: matrix.os == 'cuda'
-        run: dotnet test --logger GitHubActions --no-build --configuration=Release Src/${{ matrix.library }}.Tests.Cuda
+          dotnet-version: 5.0.x
+      
+      - name: Setup the latest .NET Core 3.1 SDK
```

Comment:
QUESTION: What is the purpose of installing the .NET Core 3.1 SDK _**and**_ the .NET 5.0 SDK, when we have pinned to .NET 5.0.x ?

Reply:
We need to install .NET Core 3.1 SDK to have the latest .NET Core 3.1 _runtime_ available for running the tests. Hence the condition for this step.

Reply:
Thanks for the clarification 👍